packages/flutter_tools/lib/src/version.dart

```diff
@@ -334,51 +389,201 @@ class FlutterVersion {
-      // Still update the timestamp to avoid us hitting the server on every single
-      // command if for some reason we cannot connect (eg. we may be offline).
-      await versionCheckStamp.store(
-        newTimeVersionWasChecked: now,
+  @override
+  final GitTagVersion gitTagVersion;
+
+  @override
+  final String frameworkVersion;
+
+  @override
+  final String channel;
+
+  @override
+  String getBranchName({bool redactUnknownBranches = false}) => channel;
+
+  @override
+  final String repositoryUrl;
+
+  @override
+  final String frameworkRevision;
+
+  @override
+  final String frameworkCommitDate;
+
+  @override
+  final String engineRevision;
+
+  @override
+  final String dartSdkVersion;
+
+  @override
+  final String devToolsVersion;
+
+  @override
+  void ensureVersionFile() {}
+}
+
+class _FlutterVersionGit extends FlutterVersion {
+  _FlutterVersionGit._({
+    required super.clock,
+    required super.flutterRoot,
+    required this.frameworkRevision,
+    required this.frameworkVersion,
+    required this.gitTagVersion,
+    required super.fs,
+  }) : super._();
+
+  @override
+  final GitTagVersion gitTagVersion;
+
+  @override
+  final String frameworkRevision;
+
+  @override
+  String get frameworkCommitDate => _gitCommitDate(lenient: true);
+
+  String? _repositoryUrl;
+  @override
+  String? get repositoryUrl {
+    final String _ = channel;
+    return _repositoryUrl;
+  }
+
+  @override
+  String get devToolsVersion => globals.cache.devToolsVersion;
+
+  @override
+  String get dartSdkVersion => globals.cache.dartSdkVersion;
+
+  @override
+  String get engineRevision => globals.cache.engineRevision;
+
+  @override
+  final String frameworkVersion;
+
+  String? _channel;
+  @override
+  String get channel {
+    String? channel = _channel;
+    if (channel == null) {
+      final String gitChannel = _runGit(
+        'git rev-parse --abbrev-ref --symbolic $kGitTrackingUpstream',
+        globals.processUtils,
+        flutterRoot,
       );
-      return null;
+      final int slash = gitChannel.indexOf('/');
+      if (slash != -1) {
+        final String remote = gitChannel.substring(0, slash);
+        _repositoryUrl = _runGit(
+          'git ls-remote --get-url $remote',
+          globals.processUtils,
+          flutterRoot,
+        );
+        channel = gitChannel.substring(slash + 1);
+      } else if (gitChannel.isEmpty) {
+        channel = 'unknown';
+      } else {
+        channel = gitChannel;
+      }
+      _channel = channel;
+    }
+    return channel;
+  }
+
+  @override
+  void ensureVersionFile() {
+    fs.file(fs.path.join(flutterRoot, 'version')).writeAsStringSync(frameworkVersion);
+    const JsonEncoder encoder = JsonEncoder.withIndent('  ');
+    final File versionFile = FlutterVersion.getVersionFile(fs, flutterRoot);
+    if (!versionFile.existsSync()) {
+      versionFile.writeAsStringSync(encoder.convert(toJson()));
```

Comment:
Will there ever be a situation that we overwrite the original `version` file while only writing the `.version.json` file when it doesn't exist?

Just curious if we should toss the original `version` inside a conditional as well?

Reply:
Good idea, done